pkg/models/vulnerability.go

```diff
@@ -167,3 +174,155 @@ func (v Vulnerability) MarshalYAML() (interface{}, error) {
+// in accordance to the Vulnerability specification.
+//
+// This is required because currently both GitHub and Pip seem to be a bit
+// inconsistent in their package name handling, so we normalize them
+// to be on the safe side.
+//
+// In the future, it's hoped that this can be improved.
+func (p Package) NormalizedName() string {
+	if p.Ecosystem != lockfile.PipEcosystem {
+		return p.Name
+	}
+
+	// per https://www.python.org/dev/peps/pep-0503/#normalized-names
+	name := regexp.MustCompile(`[-_.]+`).ReplaceAllString(p.Name, "-")
+
+	return strings.ToLower(name)
+}
+
+func (ar Range) containsVersion(pkg lockfile.PackageDetails) bool {
+	if ar.Type != RangeEcosystem && ar.Type != RangeSemVer {
+		return false
+	}
+	// todo: we should probably warn here
+	if len(ar.Events) == 0 {
+		return false
+	}
+
+	vp := semantic.MustParse(pkg.Version, pkg.CompareAs)
+
+	var affected bool
+	for _, e := range ar.Events {
+		if affected {
+			if e.Fixed != "" {
+				affected = vp.CompareStr(e.Fixed) < 0
+			} else if e.LastAffected != "" {
+				affected = e.LastAffected == pkg.Version || vp.CompareStr(e.LastAffected) <= 0
+			}
+		} else if e.Introduced != "" {
+			affected = e.Introduced == "0" || vp.CompareStr(e.Introduced) >= 0
+		}
+	}
+
+	return affected
+}
+
+type Affects []Range
+
+// affectsVersion checks if the given version is within the range
+// specified by the events of any "Ecosystem" or "Semver" type ranges
+func (a Affects) affectsVersion(pkg lockfile.PackageDetails) bool {
+	for _, r := range a {
+		if r.Type != RangeEcosystem && r.Type != RangeSemVer {
+			return false
+		}
+		if r.containsVersion(pkg) {
+			return true
+		}
+	}
+
+	return false
+}
+
+type Versions []string
+
+// MarshalJSON ensures that if there are no versions,
+// an empty array is used as the value instead of "null"
+func (vs Versions) MarshalJSON() ([]byte, error) {
+	if len(vs) == 0 {
+		return []byte("[]"), nil
+	}
+
+	out, err := json.Marshal([]string(vs))
+
+	if err != nil {
+		return out, fmt.Errorf("%w", err)
+	}
+
+	return out, nil
+}
+
+func (vs Versions) includes(v string) bool {
+	for _, v2 := range vs {
+		if v == v2 {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (v Vulnerability) isAliasOfID(id string) bool {
+	for _, alias := range v.Aliases {
+		if alias == id {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (v Vulnerability) isAliasOf(vulnerability Vulnerability) bool {
+	for _, alias := range vulnerability.Aliases {
+		if v.ID == alias || v.isAliasOfID(alias) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (v Vulnerability) AffectsEcosystem(ecosystem lockfile.Ecosystem) bool {
+	for _, affected := range v.Affected {
+		if affected.Package.Ecosystem == ecosystem {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (v Vulnerability) IsAffected(pkg lockfile.PackageDetails) bool {
+	for _, affected := range v.Affected {
+		if affected.Package.Ecosystem == pkg.Ecosystem &&
+			affected.Package.NormalizedName() == pkg.Name {
```

Comment:
Does pkg.Name also need to be Normalized?

Reply:
This can actually be removed since the osv.dev databases correctly conform to the spec - I originally had this because of https://github.com/github/advisory-database/issues/52